### PR TITLE
Replace DataLoader sampler once for IPUs

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes current IPU CI failures.

Shortens IPU CI

## Error:

We were getting the following error:

```python
        if batch_sampler is not None:
            # auto_collation with custom batch_sampler
            if batch_size != 1 or shuffle or sampler is not None or drop_last:
                raise ValueError('batch_sampler option is mutually exclusive '
                                 'with batch_size, shuffle, sampler, and '
                                 'drop_last')
            batch_size = None
            drop_last = False
        elif batch_size is None:
            # no auto_collation
            if drop_last:
>               raise ValueError('batch_size=None option disables auto-batching '
                                 'and is mutually exclusive with drop_last')
E               ValueError: batch_size=None option disables auto-batching and is mutually exclusive with drop_last

/root/miniconda3/envs/lightning/lib/python3.8/site-packages/torch/utils/data/dataloader.py:251: ValueError
```

The cause was that we were calling `replace_sampler` (which re-creates the dataloader with the correct sampler) first and `_convert_to_poptorch_loader` as a second post-process of the dataloaders which also requires recreating the dataloader.

However, in the first re-creation, a batch sampler is given which by the internal DataLoader design, sets `batch_size = None`.

```python
        if batch_sampler is not None:
            # auto_collation with custom batch_sampler
            if batch_size != 1 or shuffle or sampler is not None or drop_last:
                raise ValueError('batch_sampler option is mutually exclusive '
                                 'with batch_size, shuffle, sampler, and '
                                 'drop_last')
            batch_size = None
```

This means that on the second re-creation, we do it it with `batch_size = None` and `drop_last = True` which raises the `ValueError` observed.

To fix this, we avoid doing this double re-creation and instead monkey-patch the function so the `poptorch.DataLoader` is used directly. This required splitting `replace_sampler` into two static methods to avoid code duplication.

Even though monkey-patching like this is hacky, this solution was chosen because the IPU plugin is highly experimental and subject to change so we did not want to add an abstraction to the `Accelerator`/`TrainingTypePlugin` to support this.

Finally, some other bug fixes were required to pass the correct `RunningStage` around.

### Does your PR introduce any breaking changes? If yes, please list them.

`Accelerator` API:

- `on_reset_{train,val,test,predict}_dataloader` have been removed as they aren't necessary anymore. We don't process the dataloader twice after this PR.

`Trainer.request_dataloader` now takes a `RunningStage` instead of `str`. Not sure how public is this function.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: .azure-pipelines/ipu-tests.yml [hunk 1]
```diff
@@ -72,26 +72,11 @@ jobs:
...
       displayName: 'Testing: standard'
-
-    - bash: |
-        source ${{ variables.poplar_sdk }}/poplar-ubuntu*/enable.sh
-        source ${{ variables.poplar_sdk }}/popart-ubuntu*/enable.sh
-        export POPTORCH_WAIT_FOR_IPU=1
-        bash tests/special_tests.sh
-      env:
-        MKL_THREADING_LAYER: "GNU"
-      displayName: 'Testing: special'
```
Comment (carmocca) [inline .azure-pipelines/ipu-tests.yml:97]: There isn't any IPU special tests

File: .azure-pipelines/ipu-tests.yml [hunk 2]
```diff
@@ -72,26 +72,11 @@ jobs:
...
-        wget https://pl-public-data.s3.amazonaws.com/legacy/checkpoints.zip -P legacy/
-        unzip -o legacy/checkpoints.zip -d legacy/
-        ls -l legacy/checkpoints/
-      displayName: 'Get legacy checkpoints'
-
     - bash: |
         source ${{ variables.poplar_sdk }}/poplar-ubuntu*/enable.sh
         source ${{ variables.poplar_sdk }}/popart-ubuntu*/enable.sh
         export POPTORCH_WAIT_FOR_IPU=1
-        python -m coverage run --source pytorch_lightning -m pytest pytorch_lightning tests -v --junitxml=$(Build.StagingDirectory)/test-results.xml --durations=50
```
Comment (carmocca) [inline .azure-pipelines/ipu-tests.yml:85]: I don't think we need to run the full test suite on the IPU machine for now, here's why:

- There is only 1 IPU job available, so this is a huge bottleneck
- The IPU machine CPU is quite slow compared to the GPU one, so non-accelerator tests run considerably slower
- IPUs are experimental. By reducing the number of things this job tests, it's more feasible for us to make this a required check.

**Review by awaelchli — APPROVED** [review]

**Review by kragniz — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: .azure-pipelines/ipu-tests.yml (hunk 2 above)
Comment (kragniz) [reply .azure-pipelines/ipu-tests.yml:85]: Do we lose much IPU coverage by disabling the other tests, or are the disabled tests all run without an accelerator backend enabled?

Comment (SeanNaren) [reply .azure-pipelines/ipu-tests.yml:85]: For the actual IPU tests we lose none as all of them have been included in `pytest tests/accelerators/test_ipu.py` which is now the only file run in IPU testing after this PR is merged!

**Review by tchaton — APPROVED** [review]
LGTM !

File: .azure-pipelines/ipu-tests.yml (hunk 2 above)
Comment (kragniz) [reply .azure-pipelines/ipu-tests.yml:85]: Great!

**Pull request merged**